File: sdks/python/src/flagship/hooks/logging_hook.py
import logging
from collections.abc import Callable
from typing import TYPE_CHECKING

from openfeature.hook import Hook

if TYPE_CHECKING:
    from openfeature.flag_evaluation import FlagEvaluationDetails, FlagValueType
    from openfeature.hook import HookContext, HookHints

__all__ = ["LoggingHook"]

_default_logger = logging.getLogger("flagship").info


class LoggingHook(Hook):
    """Logs flag evaluation lifecycle events.

    Pass a custom callable to redirect output.
    """

    def __init__(self, logger: Callable[..., None] | None = None) -> None:
        self._logger = logger or _default_logger

    def before(self, hook_context: "HookContext", hints: "HookHints") -> None:
        self._logger(
            "[Flagship] Evaluating flag: %s",
            hook_context.flag_key,
        )

    def after(
        self,
        hook_context: "HookContext",
        details: "FlagEvaluationDetails[FlagValueType]",
        hints: "HookHints",
    ) -> None:
        self._logger(
            "[Flagship] Flag %s evaluated: value=%r reason=%s variant=%s",
            hook_context.flag_key,
            details.value,
            details.reason,
            details.variant,
        )

    def error(
        self,
        hook_context: "HookContext",
        exception: Exception,
        hints: "HookHints",
    ) -> None:
        self._logger(
            "[Flagship] Error evaluating flag %s: %s",
            hook_context.flag_key,
            exception,
        )

File: sdks/typescript/src/hooks/logging-hook.ts
import type { Hook, HookContext, EvaluationDetails, FlagValue, HookHints, BeforeHookContext } from '@openfeature/server-sdk';

/**
 * Logging hook for debugging flag evaluations
 *
 * @example
 * ```typescript
 * import { OpenFeature } from '@openfeature/server-sdk';
 * import { FlagshipServerProvider, LoggingHook } from '@cloudflare/flagship/server';
 *
 * const provider = new FlagshipServerProvider({ appId: 'your-app-id', accountId: 'your-account-id' });
 * await OpenFeature.setProviderAndWait(provider);
 *
 * // Add logging hook
 * OpenFeature.addHooks(new LoggingHook());
 * ```
 */
export class LoggingHook implements Hook {
	private readonly logger: (message: string, ...args: unknown[]) => void;

	constructor(logger: (message: string, ...args: unknown[]) => void = console.log) {
		this.logger = logger;
	}

	before(hookContext: BeforeHookContext, _hookHints?: HookHints): void {
		this.logger(`[Flagship] Evaluating flag: ${hookContext.flagKey}`, {
			defaultValue: hookContext.defaultValue,
			context: hookContext.context,
		});
	}

	after(hookContext: Readonly<HookContext>, evaluationDetails: EvaluationDetails<FlagValue>, _hookHints?: HookHints): void {
		this.logger(`[Flagship] Flag ${hookContext.flagKey} evaluated`, {
			value: evaluationDetails.value,
			reason: evaluationDetails.reason,
			variant: evaluationDetails.variant,
			errorCode: evaluationDetails.errorCode,
		});
	}

	error(hookContext: Readonly<HookContext>, error: unknown, _hookHints?: HookHints): void {
		const message = error instanceof Error ? error.message : String(error);
		this.logger(`[Flagship] Error evaluating flag ${hookContext.flagKey}:`, message);
	}

	finally(_hookContext: Readonly<HookContext>, _evaluationDetails: EvaluationDetails<FlagValue>, _hookHints?: HookHints): void {
		// No-op
	}
}
